scripts/generate_errors.pl

```diff
@@ -193,6 +245,110 @@
+
+sub check_range
+{
+    return if @_ <= 3;
+    my $name = shift @_;
+    my $min = shift @_;
+    my $max = pop @_;
+    foreach my $x (@_)
+    {
+        if ($x < $min || $x > $max)
+        {
+            printf STDERR ("%s (0x%04x) out of range 0x%04x-0x%04x\n",
+                           $name, $x, $min, $max);
+            exit 1;
+        }
+    }
+}
+
+my %h_low_entries;
+my %h_high_entries;
+foreach my $value (keys %error_codes_seen)
+{
+    my $name = $error_codes_seen{$value};
+    if ($value <= 0x7f)
+    {
+        $h_low_entries{$name} = {nr=>0, even=>[], odd=>[]}
+            unless exists $h_low_entries{$name};
+        ++$h_low_entries{$name}{nr};
+        push @{$h_low_entries{$name}{($value & 1 ? 'odd' : 'even')}}, $value;
+    }
+    else
+    {
+        if ($name eq 'SSL' && $value >= 0x7000)
+        {
+            # Hack because SSL occupies two high-level module IDs
+            $name = 'SSL ';
+        }
+        $h_high_entries{$name} = {nr=>0, codes=>[]}
+            unless exists $h_high_entries{$name};
+        ++$h_high_entries{$name}{nr};
+        push @{$h_high_entries{$name}{codes}}, $value;
+    }
+}
+foreach my $name (keys %h_low_entries)
+{
+    my $entry = $h_low_entries{$name};
+    my @even = sort {$a <=> $b} @{$entry->{even}};
+    my @odd = sort {$a <=> $b} @{$entry->{odd}};
+    check_range($name, @even);
+    check_range($name, @odd);
+    $entry->{ranges} = (@even ?
+                        sprintf("0x%04X-0x%04X", $even[0], $even[@even-1]) :
+                        "             ");
+    $entry->{ranges} .= sprintf("   0x%04X-0x%04X", $odd[0], $odd[@odd-1])
+        if @odd;
+    $entry->{sort_key} = $entry->{ranges};
+    $entry->{sort_key} =~ s/\A /~/;
+}
+foreach my $name (keys %h_high_entries)
+{
+    my $entry = $h_high_entries{$name};
+    my @codes = sort {$a <=> $b} @{$entry->{codes}};
+    check_range($name, @codes);
+    $entry->{id} = $codes[@codes-1] >> 12;
+    $entry->{comment} = (($codes[0] & 0xf80) == 0x080 ? '' :
+                         ($codes[@codes-1] & 0xf80) == 0xf80 ?
+                         ' (Started from top)' :
+                         ' (Started from middle)');
+    $entry->{sort_key} = sprintf("%04x", $codes[0]);
+}
+my $h_low_text =
+    join('',
+         map {sprintf(" * %-9s %2d  %s\n", $_,
+                      $h_low_entries{$_}{nr},
+                      $h_low_entries{$_}{ranges})}
+         sort {$h_low_entries{$a}{sort_key} cmp $h_low_entries{$b}{sort_key}}
+         keys %h_low_entries);
+my $h_high_text =
+    join('',
+         map {sprintf(" * %-9s %2d  %d%s\n", $_,
+                      $h_high_entries{$_}{id},
+                      $h_high_entries{$_}{nr},
+                      $h_high_entries{$_}{comment})}
+         sort {$h_high_entries{$a}{sort_key} cmp $h_high_entries{$b}{sort_key}}
+         keys %h_high_entries);
+
+open(ERROR_FILE, "+<$error_h") or die "Opening destination file '$error_h': $!";
```

Comment:
I quite like the way `error.h` is updated in place, but this is not consistent with the way other files are generated (from a template in `scripts/data_files`. I feel like this discrepancy should be clearly documented in the comments at the beginning of the script. (In the long run, I'd be in favour of resolving the discrepancy by using this method for other generated files as well, but I understand this might go a bit beyond the scope of this PR, so just documenting it in the meantime is fine.)